src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+    }
+  }
+
+  UMPIRE_LOG(Debug, "Using chunk " << chunk << " with data " << chunk->data << " and size " << chunk->size
+                                   << " for allocation of size " << rounded_bytes);
+
+  if ((chunk->size == chunk->chunk_size) && chunk->free) {
+    m_releasable_bytes -= chunk->chunk_size;
+    m_releasable_blocks--;
+  }
+
+  void* ret = chunk->data;
+  m_used_map.insert(std::make_pair(ret, chunk));
+
+  chunk->free = false;
+
+  if (rounded_bytes != chunk->size) {
+    std::size_t remaining{chunk->size - rounded_bytes};
+    UMPIRE_LOG(Debug, "Splitting chunk " << chunk->size << "into " << rounded_bytes << " and " << remaining);
+
+    void* chunk_storage{m_chunk_pool.allocate()};
+    Chunk* split_chunk{new (chunk_storage)
+                           Chunk{static_cast<char*>(ret) + rounded_bytes, remaining, chunk->chunk_size, r}};
+    auto old_next = chunk->next;
+    chunk->next = split_chunk;
+    split_chunk->prev = chunk;
+    split_chunk->next = old_next;
+
+    if (split_chunk->next)
+      split_chunk->next->prev = split_chunk;
+
+    chunk->size = rounded_bytes;
+    split_chunk->size_map_it = m_free_map.insert(std::make_pair(remaining, split_chunk));
+    split_chunk->free = true;
+  }
+
+  m_current_bytes += rounded_bytes;
+
+  UMPIRE_UNPOISON_MEMORY_REGION(m_allocator, ret, bytes);
+  return ret;
+}
+
+void ResourceAwarePool::deallocate(void* ptr, std::size_t size)
+{
+  auto r = getResource(ptr);
+
+  UMPIRE_LOG(Warning, fmt::format("The ResourceAwarePool requires a Camp resource. Calling deallocate with: {}.",
+                                  camp::resources::to_string(r)));
+
+  deallocate_resource(r, ptr, size);
+}
+
+void ResourceAwarePool::do_deallocate(Chunk* chunk, void* ptr) noexcept
+{
+  UMPIRE_POISON_MEMORY_REGION(m_allocator, ptr, chunk->size);
+  UMPIRE_USE_VAR(ptr);
+  chunk->free = true;
+
+  // Removing chunk from pending
+  for (auto it = m_pending_map.begin(); it != m_pending_map.end();) {
+    auto my_chunk = (*it);
+    if (my_chunk == chunk) {
+      m_pending_map.erase(it);
+    } else {
+      it++;
+    }
+  }
+
+  UMPIRE_LOG(Debug, "In the do_deallocate function. Deallocating data held by " << chunk);
+
+  if (chunk->prev && chunk->prev->free == true) {
+    auto prev = chunk->prev;
+
+    UMPIRE_LOG(Debug, "Removing chunk" << prev << " from free map");
+    m_free_map.erase(prev->size_map_it);
+
+    prev->size += chunk->size;
+    prev->next = chunk->next;
+
+    prev->m_event = chunk->m_event;
+    prev->m_resource = chunk->m_resource;
+
+    if (prev->next)
+      prev->next->prev = prev;
+
+    UMPIRE_LOG(Debug, "Merging with prev" << prev << " and " << chunk);
+    UMPIRE_LOG(Debug, "New size: " << prev->size);
+
+    chunk->~Chunk(); // manually call destructor
+    m_chunk_pool.deallocate(chunk);
+    chunk = prev;
+  }
+
+  if (chunk->next && chunk->next->free == true) {
+    auto next = chunk->next;
+
+    UMPIRE_LOG(Debug, "Removing chunk" << next << " from free map");
```

Comment:
It might be worth making a merge_chunk function since the code is basically identical for merging with both the previous and the next chunks.

Reply:
This would be the subject of a different PR - this sort of implementation is what is currently done in QuickPool too. It'd be a separate PR on making that merging function. @davidbeckingsale any preference on reworking that?

Reply:
I created a separate issue for it: https://github.com/LLNL/Umpire/issues/921

Reply:
You could always do a PR to merge that into this PR.